numpy/_core/src/umath/matmul.c.src

```diff
@@ -499,9 +525,46 @@ NPY_NO_EXPORT void
                  * non-blasable (or non-ccontiguous output)
                  * we could still use BLAS, see gh-12365.
                  */
-                @TYPE@_matmul_inner_noblas(ip1, is1_m, is1_n,
-                                           ip2, is2_n, is2_p,
-                                           op, os_m, os_p, dm, dn, dp);
+                if (tmp_ip12op == NULL) {
+                    npy_intp ip1_size = sz * dm * dn,
+                             ip2_size = sz * dn * dp,
+                             op_size = sz * dm * dp,
+                             total_size = ip1_size + ip2_size + op_size;
+
+                    tmp_ip12op = (uint8_t*)malloc(total_size);
+
+                    tmp_ip1 = tmp_ip12op;
```

Comment:
There is still @seberg's old comment here, that we should only copy if needed. E.g., the output almost certainly is in the right order, so one would get something like
```
if (o_c_blasable) {
    tmp_op = op;
}
else {
    tmp_op = (uint8_t*)...
}
```
and then further down
```
if (tmp_op != op) {
    @TYPE@_matrix_copy(...);
}
```

Reply:
Yeah, I've got that in my pipeline already. I think though from a compiler perspective it would make more sense to check for `i1blasable` (in my case) twice. Easier to optimize.

Reply:
I like the way you approached this!

Reply:
I was thinking we might want to swap strides to ensure that these perform identically in the copy (and thus overall, since I don't think order will matter in blas):
```
a = np.ones((100, 1000, 1000))[:, ::10, ::10]
b = np.ones((100, 1000, 1000))[:, ::10, ::10]

%timeit res = a @ b
%timeit res = a.mT @ b.mT
```
But my first timings show at most a 50% slowdown for something like that.

One can probably construct some worst case scenario (and it may be worse on other hardware), but that is not significant enough that I care if you prefer not doing it here or at all.

Reply:
Okay, I think now I understand what you mean. Basically, when copying, we want to both iterate the smaller of `is_n` and `is_m` in the inner loop and the smaller one of `os_n` and `os_m` in the inner loop and declare the thing transposed if we can't have both at the same time to let BLAS handle it.

Reply:
Right, that could speed up the copy while it should make no difference for the matmul.  But... I am starting to see more clearly that the copy tends to be fast by comparison, so it seems to me its not a big deal here.

Reply:
Too bad, just "fixed" it 😅 I can also revert it if that helps. BTW which BLAS are you using? OpenBLAS or BLIS? Could be OpenBLAS related, BLIS makes a more robust impression on me.